pkg/sources/nats/nats_test.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nats
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/nats-io/nats-server/v2/server"
+	natstestserver "github.com/nats-io/nats-server/v2/test"
+	natslib "github.com/nats-io/nats.go"
+	"github.com/stretchr/testify/assert"
+
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/isb/stores/simplebuffer"
+	"github.com/numaproj/numaflow/pkg/watermark/generic"
+	"github.com/numaproj/numaflow/pkg/watermark/store"
+	"github.com/numaproj/numaflow/pkg/watermark/store/noop"
+)
+
+func testVertex(t *testing.T, url, subject, queue string, hostname string, replicaIndex int32) *dfv1.VertexInstance {
+	t.Helper()
+	v := &dfv1.Vertex{
+		Spec: dfv1.VertexSpec{
+			AbstractVertex: dfv1.AbstractVertex{
+				Name: "test-v",
+				Source: &dfv1.Source{
+					Nats: &dfv1.NatsSource{
+						URL:     url,
+						Subject: subject,
+						Queue:   queue,
+					},
+				},
+			},
+		},
+	}
+	vi := &dfv1.VertexInstance{
+		Vertex:   v,
+		Hostname: hostname,
+		Replica:  replicaIndex,
+	}
+	return vi
+}
+
+func newInstance(t *testing.T, vi *dfv1.VertexInstance) (*natsSource, error) {
+	t.Helper()
+	dest := []isb.BufferWriter{simplebuffer.NewInMemoryBuffer("test", 100)}
+	publishWMStores := store.BuildWatermarkStore(noop.NewKVNoOpStore(), noop.NewKVNoOpStore())
+	fetchWatermark, publishWatermark := generic.BuildNoOpWatermarkProgressorsFromBufferMap(map[string]isb.BufferWriter{})
+	return New(vi, dest, fetchWatermark, publishWatermark, publishWMStores, WithReadTimeout(1*time.Second))
+}
+
+func runNatsServer(t *testing.T) *server.Server {
+	t.Helper()
+	opts := natstestserver.DefaultTestOptions
+	opts.Port = 9989
+	opts.Cluster.Name = "testing"
+	return natstestserver.RunServer(&opts)
```

Comment:
this is cool, do they have a JS version too?

Reply:
Same thing can be used from JS.